### PR TITLE
Use call for string to numeric conversion

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes issue #10594.

### CODE REVIEW

File: base/deprecated.jl [hunk 1]
```diff
@@ -409,12 +395,7 @@ for (f,t) in ((:int,    Int), (:int8,   Int8), (:int16,  Int16), (:int32,  Int32
               (:int64,  Int64), (:int128, Int128), (:uint,   UInt), (:uint8,  UInt8),
               (:uint16, UInt16), (:uint32, UInt32), (:uint64, UInt64), (:uint128,UInt128))
```
Comment (tkelman) [inline base/deprecated.jl]: should this list get Float32 and Float64 added to it?

File: base/deprecated.jl [hunk 2]
```diff
@@ -296,22 +296,8 @@ const MemoryError = OutOfMemoryError
...
-@deprecate int16(s::AbstractString)     parseint(Int16,s)
-@deprecate uint16(s::AbstractString)    parseint(UInt16,s)
-@deprecate int32(s::AbstractString)     parseint(Int32,s)
-@deprecate uint32(s::AbstractString)    parseint(UInt32,s)
-@deprecate int64(s::AbstractString)     parseint(Int64,s)
-@deprecate uint64(s::AbstractString)    parseint(UInt64,s)
-@deprecate int128(s::AbstractString)    parseint(Int128,s)
-@deprecate uint128(s::AbstractString)   parseint(UInt128,s)
-@deprecate float64(s::AbstractString)   parsefloat(Float64,s)
-@deprecate float32(s::AbstractString)   parsefloat(Float32,s)
```
Comment (tkelman) [inline base/deprecated.jl]: aren't these deprecations still necessary?

Comment (simonbyrne) [inline base/deprecated.jl]: These are already handled by the generic method (https://github.com/JuliaLang/julia/blob/sb/parsefloat/base/deprecated.jl#L360-L377)

Comment (tkelman) [inline base/deprecated.jl]: whoops, yeah, hiding not-in-the-diff

File: base/deprecated.jl (hunk 1 above)
Comment (simonbyrne) [inline base/deprecated.jl]: These should be handled by https://github.com/JuliaLang/julia/blob/sb/parsefloat/base/deprecated.jl#L312-L317

**Comment by garborg** [comment]
I don't want to rush any decisions, but this would be great to resolve one way or the other before packages like `JuliaParser.jl` and [`LibExpat.jl`](https://github.com/amitmurthy/LibExpat.jl/pull/25) make decisions about whether to keep exporting a `parse` that doesn't extend `Base.parse`.

**Comment by JeffBezanson** [comment]
I guess this is ok. IIRC, our thinking used to be that for something as simple as a number, construction and conversion should be the same thing, and that parsing should be done explicitly since it is slow.

**Comment by stevengj** [comment]
Looks good to me.

**Comment by stevengj** [comment]
In this case, why do we need `parse{T<:Number}(T, x)`?

**Comment by JeffBezanson** [comment]
Yes, I think that's a bit of a problem. If we have both, it's not clear which one to call or define. If we use only constructors, then the interface to parsing is construction. However this doesn't make sense for all types. For example `List("(1 2)")` could construct a list of one string, but `parse(List, "(1 2)")` would construct a list of two integers.

**Pull request closed without merging**